components/remote_settings/src/client.rs

```diff
@@ -48,14 +59,59 @@ impl Client {
             bucket_name,
             collection_name: config.collection_name,
             remote_state: Default::default(),
+            cache,
         })
     }
 
     /// Fetches all records for a collection that can be found in the server,
     /// bucket, and collection defined by the [ClientConfig] used to generate
     /// this [Client].
     pub fn get_records(&self) -> Result<RemoteSettingsResponse> {
-        self.get_records_with_options(&GetItemsOptions::new())
+        match self.cache.as_deref() {
+            None => self.get_records_with_options(&GetItemsOptions::new()),
+            Some(cache) => self.get_records_with_cache(cache),
+        }
+    }
+
+    /// Fetches records, using a cached response for efficiency.
+    fn get_records_with_cache(
+        &self,
+        cache: &dyn RemoteSettingsCache,
+    ) -> Result<RemoteSettingsResponse> {
+        let cached_response = cache.get();
+        let cached_last_modified = cached_response
+            .as_ref()
+            .map(|r| r.last_modified)
+            .unwrap_or(0);
+
+        let response = match cached_response {
+            None => self.get_records()?,
+            Some(cached) => {
+                let new = self.get_records_since(cached.last_modified)?;
+                cache::merge_cache_and_response(cached, new)
+            }
+        };
+        if cached_last_modified != response.last_modified {
+            cache.store(response.clone());
+        }
+        Ok(response)
+    }
+
+    pub fn get_cached_records(&self) -> Option<RemoteSettingsResponse> {
+        self.cache.as_deref().and_then(|cache| cache.get())
```

Comment:
None here seems mildly odd - at face value I'd have thought calling this (and below) with self.cache is a logic error of some kind, and subtly different from "cache exists but does not have it".

Or maybe not! I'm fine with this if you think it really is fine and I can see how some parts of the code really don't care about the distinction, but thought I'd throw it out anyway.

Reply:
I can go either way as well.  The thing I like about this is that it presents a simpler API for consumers who do specify a cache (not throwing at all is simpler than telling users that throwing won't happen in practice).